clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -157,6 +158,99 @@ class Util {
 
 } // anonymous namespace
 
+ExprResult Sema::ActOnSYCLBuiltinNumFieldsExpr(ParsedType PT) {
+  TypeSourceInfo *TInfo = nullptr;
+  QualType QT = GetTypeFromParser(PT, &TInfo);
+  assert(TInfo && "couldn't get type info from a type from the parser?");
+  SourceLocation TypeLoc = TInfo->getTypeLoc().getBeginLoc();
+
+  return BuildSYCLBuiltinNumFieldsExpr(TypeLoc, QT);
+}
+
+ExprResult Sema::BuildSYCLBuiltinNumFieldsExpr(SourceLocation Loc,
+                                               QualType SourceTy) {
+  int64_t NumFields = 0;
+  if (!SourceTy->isDependentType()) {
+    if (RequireCompleteType(Loc, SourceTy,
+                            diag::err_sycl_type_trait_requires_complete_type,
+                            /*__builtin_num_fields*/ 0))
+      return ExprError();
+
+    if (!SourceTy->isRecordType()) {
+      Diag(Loc, diag::err_sycl_type_trait_requires_record_type)
+          << /*__builtin_num_fields*/ 0;
+      return ExprError();
+    }
+
+    RecordDecl *RD = SourceTy->getAsRecordDecl();
+    assert(RD && "Record type but no record decl?");
+
+    NumFields = std::distance(RD->field_begin(), RD->field_end());
```

Comment:
It is a touch sad that we're storing this info 2x.  We Need the SourceTy in case it is dependent, but just looking up NumFields during evaluation seems like less work.